### PR TITLE
[AlwaysOn] Add instrumented tests

### PR DESCRIPTION
**Pull request by alexvanyo**
Adds some basic instrumented tests for the `AlwaysOn` sample, checking that the views are updated in both ambient and active modes.

### CODE REVIEW

File: AlwaysOn/Wearable/src/androidTest/java/com/example/android/wearable/wear/alwayson/MainActivityTests.kt [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+    private val context = ApplicationProvider.getApplicationContext<Application>()
+
+    private val testDispatcher: TestCoroutineDispatcher = TestCoroutineDispatcher()
+
+    /**
+     * A timestamp in the relatively far, far future (year 2200).
+     *
+     * This ensures the real alarm manager won't actually trigger.
+     */
+    private var instant = Instant.ofEpochMilli(7258118400000)
```
Comment (codingjeremy) [inline AlwaysOn/Wearable/src/androidTest/java/com/example/android/wearable/wear/alwayson/MainActivityTests.kt:58]: Could you make this a constant so it's easier to read, e.g.,

private const val YEAR_2200 = 7258118400000

File: AlwaysOn/Wearable/src/androidTest/java/com/example/android/wearable/wear/alwayson/MainActivityTests.kt [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    fun teardown() {
+        scenario.close()
+    }
+
+    @Test
+    fun initialTextIsCorrect(): Unit = runBlocking {
+        scenario.moveToState(Lifecycle.State.RESUMED)
+
+        onView(withId(R.id.time)).check(matches(withText("00:00:00")))
+        onView(withId(R.id.time_stamp)).check(matches(withText(context.getString(R.string.timestamp_label, 7258118400000L))))
```
Comment (codingjeremy) [inline AlwaysOn/Wearable/src/androidTest/java/com/example/android/wearable/wear/alwayson/MainActivityTests.kt:89]: Could you use the same constant here (convert to Long)?

File: AlwaysOn/Wearable/src/androidTest/java/com/example/android/wearable/wear/alwayson/MainActivityTests.kt [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    @Test
+    fun textIsCorrectAfterFiveSeconds(): Unit = runBlocking {
+        scenario.moveToState(Lifecycle.State.RESUMED)
+
+        // Advance 5 seconds, one at a time
+        repeat(5) {
+            advanceTime(Duration.ofSeconds(1))
+        }
+
+        onView(withId(R.id.time)).check(matches(withText("00:00:05")))
```
Comment (codingjeremy) [inline AlwaysOn/Wearable/src/androidTest/java/com/example/android/wearable/wear/alwayson/MainActivityTests.kt:103]: Could you make this a constant? You use it other locations too.

File: AlwaysOn/Wearable/src/androidTest/java/com/example/android/wearable/wear/alwayson/MainActivityTests.kt [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+    @After
+    fun teardown() {
+        scenario.close()
+    }
+
+    @Test
+    fun initialTextIsCorrect(): Unit = runBlocking {
+        scenario.moveToState(Lifecycle.State.RESUMED)
+
+        onView(withId(R.id.time)).check(matches(withText("00:00:00")))
```
Comment (codingjeremy) [inline AlwaysOn/Wearable/src/androidTest/java/com/example/android/wearable/wear/alwayson/MainActivityTests.kt:88]: If you make 00:00:05 a constant you should do it here too.

File: AlwaysOn/Wearable/src/androidTest/java/com/example/android/wearable/wear/alwayson/MainActivityTests.kt [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+        repeat(5) {
+            advanceTime(Duration.ofSeconds(1))
+        }
+
+        uiDevice.pressKeyCode(KeyEvent.KEYCODE_SLEEP)
+        Thread.sleep(1000) // Ugly sleep, without it sometimes ambient mode won't be entered
+        Espresso.onIdle()
+
+        onView(withId(R.id.time)).check(matches(withText("00:00:05")))
+        onView(withId(R.id.time_stamp)).check(matches(withText(context.getString(R.string.timestamp_label, 7258118405000L))))
```
Comment (codingjeremy) [inline AlwaysOn/Wearable/src/androidTest/java/com/example/android/wearable/wear/alwayson/MainActivityTests.kt:123]: use constant

File: AlwaysOn/Wearable/src/androidTest/java/com/example/android/wearable/wear/alwayson/MainActivityTests.kt [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+        scenario.onActivity {
+            PendingIntent.getBroadcast(
+                it, 0, Intent(MainActivity.AMBIENT_UPDATE_ACTION), PendingIntent.FLAG_UPDATE_CURRENT
+            ).send()
+        }
+
+        Espresso.onIdle()
+
+        onView(withId(R.id.time)).check(matches(withText("00:00:10")))
+        onView(withId(R.id.time_stamp)).check(matches(withText(context.getString(R.string.timestamp_label, 7258118410500L))))
```
Comment (codingjeremy) [inline AlwaysOn/Wearable/src/androidTest/java/com/example/android/wearable/wear/alwayson/MainActivityTests.kt:154]: Constant

File: AlwaysOn/Wearable/src/androidTest/java/com/example/android/wearable/wear/alwayson/MainActivityTests.kt [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+        advanceTime(Duration.ofSeconds(5))
+        scenario.onActivity {
+            PendingIntent.getBroadcast(
+                it, 0, Intent(MainActivity.AMBIENT_UPDATE_ACTION), PendingIntent.FLAG_UPDATE_CURRENT
+            ).send()
+        }
+
+        Espresso.onIdle()
+
+        onView(withId(R.id.time)).check(matches(withText("00:00:10")))
```
Comment (codingjeremy) [inline AlwaysOn/Wearable/src/androidTest/java/com/example/android/wearable/wear/alwayson/MainActivityTests.kt:153]: Constant

File: AlwaysOn/Wearable/src/androidTest/java/com/example/android/wearable/wear/alwayson/MainActivityTests.kt [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+
+        advanceTime(Duration.ofSeconds(2))
+
+        // Exit ambient mode
+        uiDevice.pressKeyCode(KeyEvent.KEYCODE_WAKEUP)
+        Thread.sleep(1000) // Ugly sleep, without it sometimes ambient mode won't be exited
+        Espresso.onIdle()
+
+        onView(withId(R.id.time)).check(matches(withText("00:00:12")))
+        onView(withId(R.id.time_stamp)).check(matches(withText(context.getString(R.string.timestamp_label, 7258118412500L))))
```
Comment (codingjeremy) [inline AlwaysOn/Wearable/src/androidTest/java/com/example/android/wearable/wear/alwayson/MainActivityTests.kt:193]: Same here.

File: AlwaysOn/Wearable/src/androidTest/java/com/example/android/wearable/wear/alwayson/MainActivityTests.kt [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+        Espresso.onIdle()
+
+        advanceTime(Duration.ofSeconds(2))
+
+        // Exit ambient mode
+        uiDevice.pressKeyCode(KeyEvent.KEYCODE_WAKEUP)
+        Thread.sleep(1000) // Ugly sleep, without it sometimes ambient mode won't be exited
+        Espresso.onIdle()
+
+        onView(withId(R.id.time)).check(matches(withText("00:00:12")))
```
Comment (codingjeremy) [inline AlwaysOn/Wearable/src/androidTest/java/com/example/android/wearable/wear/alwayson/MainActivityTests.kt:192]: Same here.

**Review by codingjeremy — COMMENTED** [review]
Just some minor nits.

File: AlwaysOn/Wearable/src/androidTest/java/com/example/android/wearable/wear/alwayson/MainActivityTests.kt (hunk 1 above)
Comment (alexvanyo) [reply AlwaysOn/Wearable/src/androidTest/java/com/example/android/wearable/wear/alwayson/MainActivityTests.kt]: Updated! I only defined one timestamp, and just added an offset for the other usages.

File: AlwaysOn/Wearable/src/androidTest/java/com/example/android/wearable/wear/alwayson/MainActivityTests.kt (hunk 3 above)
Comment (alexvanyo) [reply AlwaysOn/Wearable/src/androidTest/java/com/example/android/wearable/wear/alwayson/MainActivityTests.kt]: Done, I added constants for each of the time displays.

File: AlwaysOn/Wearable/src/androidTest/java/com/example/android/wearable/wear/alwayson/MainActivityTests.kt [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+        scenario.onActivity {
+            PendingIntent.getBroadcast(
+                it, 0, Intent(MainActivity.AMBIENT_UPDATE_ACTION), PendingIntent.FLAG_UPDATE_CURRENT
+            ).send()
+        }
+
+        Espresso.onIdle()
+
+        onView(withId(R.id.time)).check(matches(withText(TEN_SEC_DISPLAY)))
+        onView(withId(R.id.time_stamp)).check(matches(withText(context.getString(R.string.timestamp_label, YEAR_2200 + 10500L))))
```
Comment (codingjeremy) [inline AlwaysOn/Wearable/src/androidTest/java/com/example/android/wearable/wear/alwayson/MainActivityTests.kt:154]: These offsets would be easier to read as constants too. :)

It's hard for me to figure out the value just form glancing at it vs. TEN_SECONDS

File: AlwaysOn/Wearable/src/androidTest/java/com/example/android/wearable/wear/alwayson/MainActivityTests.kt [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+
+        advanceTime(Duration.ofSeconds(2))
+
+        // Exit ambient mode
+        uiDevice.pressKeyCode(KeyEvent.KEYCODE_WAKEUP)
+        Thread.sleep(1000) // Ugly sleep, without it sometimes ambient mode won't be exited
+        Espresso.onIdle()
+
+        onView(withId(R.id.time)).check(matches(withText(TWELVE_SEC_DISPLAY)))
+        onView(withId(R.id.time_stamp)).check(matches(withText(context.getString(R.string.timestamp_label, YEAR_2200 + 12500L))))
```
Comment (codingjeremy) [inline AlwaysOn/Wearable/src/androidTest/java/com/example/android/wearable/wear/alwayson/MainActivityTests.kt:193]: Same here

**Review by codingjeremy — COMMENTED** [review]
Sorry, just a couple last nits. BTW, I would add new commits to the PR, not append to the same commit... otherwise it's a little confusing to read in the PR... the PR will still show it nice and different... i.e., I would avoid force push. 

Thanks for the work!

File: AlwaysOn/Wearable/src/androidTest/java/com/example/android/wearable/wear/alwayson/MainActivityTests.kt (hunk 10 above)
Comment (alexvanyo) [reply AlwaysOn/Wearable/src/androidTest/java/com/example/android/wearable/wear/alwayson/MainActivityTests.kt]: Updated to use `Duration` math here, to get the self-documenting `plusSeconds` and `plusMillis`.

**Review by codingjeremy — APPROVED** [review]
Awesome, LGTM!

**Pull request merged**